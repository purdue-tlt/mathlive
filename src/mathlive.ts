/* eslint-disable no-new */
import type { Mathfield } from './public/mathfield';
import type {
  MathfieldOptions,
  RemoteVirtualKeyboardOptions,
  TextToSpeechOptions,
} from './public/options';
import type {
  ErrorListener,
  ParserErrorCode,
  MathfieldErrorCode,
} from './public/core';

import { Atom } from './core/atom-class';
import { makeRoot } from './core/atom-utils';
import { parseLatex } from './core/parser';
import { coalesce, makeStruts, Span } from './core/span';
import { MACROS, MacroDictionary } from './core-definitions/definitions';
import { MathfieldPrivate } from './editor-mathfield/mathfield-private';
import AutoRender, { AutoRenderOptionsPrivate } from './addons/auto-render';
import {
  MathJsonLatexOptions,
  MathJson,
  atomtoMathJson,
  jsonToLatex,
} from './addons/math-json';
import MathLiveDebug, {
  asciiMathToLatex,
  latexToAsciiMath,
} from './addons/debug';
import { MATHSTYLES } from './core/mathstyle';
import { defaultSpeakHook } from './editor/speech';
import {
  defaultReadAloudHook,
  readAloudStatus,
  pauseReadAloud,
  resumeReadAloud,
  playReadAloud,
} from './editor/speech-read-aloud';
import { atomToSpeakableText } from './editor/atom-to-speakable-text';
import { atomsToMathML } from './addons/math-ml';

import './addons/definitions-metadata';

import './editor/virtual-keyboard-commands';
import { RemoteVirtualKeyboard } from './editor-mathfield/remote-virtual-keyboard';

export { MathfieldElement } from './public/mathfield-element';

export { makeRoot };

export function makeMathField(
  element: HTMLElement,
  options: Partial<MathfieldOptions> = {}
): Mathfield {
  options.speakHook = options.speakHook ?? defaultSpeakHook;
  options.readAloudHook = options.readAloudHook ?? defaultReadAloudHook;
  return new MathfieldPrivate(getElement(element), options);
}

export function makeSharedVirtualKeyboard(
  options: Partial<RemoteVirtualKeyboardOptions>
): void {
  new RemoteVirtualKeyboard(options);
}

export function convertLatexToMathlist(
  latex: string,
  options?: {
    macros?: MacroDictionary;
    onError?: ErrorListener<ParserErrorCode>;
  }
): Atom[] {
  options = options ?? {};
  options.macros = options.macros ?? {};
  options.macros = { ...MACROS, ...(options.macros ?? {}) };

  const mathlist = parseLatex(
    latex,
    'math',
    null,
    options.macros,
    false,
    options.onError
  );

  return mathlist;
}

/** @deprecated */
function latexToMarkup(
  text: string,
  options?: {
    mathstyle?: 'displaystyle' | 'textstyle';
    letterShapeStyle?: 'tex' | 'french' | 'iso' | 'upright' | 'auto';
    macros?: MacroDictionary;
    onError?: ErrorListener<ParserErrorCode>;
    format?: string;
  }
): string {
  return convertLatexToMarkup(text, options);
}

export function convertLatexToMarkup(
  text: string,
  options?: {
    mathstyle?: 'displaystyle' | 'textstyle';
    letterShapeStyle?: 'tex' | 'french' | 'iso' | 'upright' | 'auto';
    macros?: MacroDictionary;
    onError?: ErrorListener<ParserErrorCode>;
    format?: string;
    onCreateMathlist?: (mathlist: Atom[]) => void;
  }
): string {
  options = options ?? {};
  options.mathstyle = options.mathstyle || 'displaystyle';
  options.letterShapeStyle = options.letterShapeStyle || 'auto';
  options.macros = { ...MACROS, ...(options.macros ?? {}) };

  //
  // 1. Parse the formula and return a tree of atoms, e.g. 'genfrac'.
  //

  const atoms = parseLatex(
    text,
    'math',
    null,
    options.macros,
    false,
    options.onError
  );

  if (options.onCreateMathlist) {
    options.onCreateMathlist(atoms);
  }

  //
  // 2. Transform the math atoms into elementary spans
  //    for example from genfrac to vlist.
  //
  let spans = Atom.render(
    {
      mathstyle: MATHSTYLES[options.mathstyle],
      letterShapeStyle: options.letterShapeStyle,
    },
    atoms
  );

  //
  // 3. Simplify by coalescing adjacent nodes
  //    for example, from <span>1</span><span>2</span>
  //    to <span>12</span>
  //
  spans = coalesce(spans);

  //
  // 4. Wrap the expression with struts
  //
  const wrapper = makeStruts(new Span(spans, 'ML__base'), 'ML__mathlive');

  //
  // 5. Generate markup
  //

  return wrapper.toMarkup({ hscale: 1 });
}

export function convertLatexToMathMl(
  latex: string,
  options: Partial<{
    macros: MacroDictionary;
    onError: ErrorListener<ParserErrorCode>;
    generateID: boolean;
  }> = {}
): string {
  options.macros = { ...MACROS, ...(options.macros ?? {}) };

  return atomsToMathML(
    parseLatex(latex, 'math', [], options.macros, false, options.onError),
    options
  );
}

/** @deprecated */
function latexToMathML(
  latex: string,
  options?: Partial<{
    macros: MacroDictionary;
    onError: ErrorListener<ParserErrorCode>;
    generateID: boolean;
  }>
): string {
  return convertLatexToMathMl(latex, options);
}

/** @deprecated Use MathJSON */
function latexToAST(
  latex: string,
  options?: MathJsonLatexOptions & {
    macros?: MacroDictionary;
    onError?: ErrorListener<ParserErrorCode | string>;
  }
): MathJson {
  options = options ?? {};
  options.macros = { ...MACROS, ...(options.macros ?? {}) };

  // Return parseLatex(latex, options);

  return atomtoMathJson(
    parseLatex(latex, 'math', null, options.macros, false, options.onError),
    options
  );
}

/** @deprecated Use MathJSON */
export function astToLatex(
  expr: MathJson,
  options: MathJsonLatexOptions
): string {
  return jsonToLatex(
    typeof expr === 'string' ? JSON.parse(expr) : expr,
    options
  );
  // Return emitLatex(expr, options);
}

export function convertLatexToSpeakableText(
  latex: string,
  options: Partial<
    TextToSpeechOptions & {
      macros?: MacroDictionary;
      onError?: ErrorListener<ParserErrorCode | MathfieldErrorCode>;
    }
  > = {}
): string {
  options.macros = options.macros ?? {};
  Object.assign(options.macros, MACROS);

  const mathlist = parseLatex(
    latex,
    'math',
    null,
    options.macros,
    false,
    options.onError
  );

  return atomToSpeakableText(
    mathlist,
    options as Required<TextToSpeechOptions>
  );
}

/** @deprecated */
function latexToSpeakableText(
  latex: string,
  options?: Partial<
    TextToSpeechOptions & {
      macros?: MacroDictionary;
      onError?: ErrorListener<ParserErrorCode | MathfieldErrorCode>;
    }
  >
): string {
  return convertLatexToSpeakableText(latex, options);
}

export function renderMathInDocument(options: AutoRenderOptionsPrivate): void {
  renderMathInElement(document.body, options);
}

function getElement(element: string | HTMLElement): HTMLElement {
  if (typeof element === 'string') {
    const result: HTMLElement = document.getElementById(element);
    if (result === null) {
      throw new Error(`The element with ID "${element}" could not be found.`);
    }

    return result;
  }

  return element;
}

export function renderMathInElement(
  element: HTMLElement,
  options: AutoRenderOptionsPrivate
): void {
  options = options ?? {};
  options.renderToMarkup = options.renderToMarkup ?? convertLatexToMarkup;
  options.renderToMathML = options.renderToMathML ?? convertLatexToMathMl;
  options.renderToSpeakableText =
    options.renderToSpeakableText ?? convertLatexToSpeakableText;
  options.macros = options.macros ?? MACROS;
  AutoRender.renderMathInElement(getElement(element), options);
}

function validateNamespace(options): void {
  if (typeof options.namespace === 'string') {
    if (!/^[a-z]+-?$/.test(options.namespace)) {
      throw new Error(
        'options.namespace must be a string of lowercase characters only'
      );
    }

    if (!options.namespace.endsWith('-')) {
      options.namespace += '-';
    }
  }
}

/** @deprecated */
function revertToOriginalContent(
  element: string | HTMLElement,
  options: AutoRenderOptionsPrivate
): void {
  deprecatedDefaultImport('revertToOriginalContent');
  //  If (element instanceof MathfieldPrivate) {
  //      element.$revertToOriginalContent();
  //    } else {
  // element is a pair: accessible span, math -- set it to the math part
  element = getElement(element).children[1] as HTMLElement;
  options = options ?? {};
  validateNamespace(options);
  const html = element.getAttribute(
    'data-' + (options.namespace ?? '') + 'original-content'
  );
  element.innerHTML =
    typeof options.createHTML === 'function' ? options.createHTML(html) : html;
  //  }
}

/** @deprecated */
function getOriginalContent(
  element: string | HTMLElement,
  options: AutoRenderOptionsPrivate
): string {
  deprecatedDefaultImport('getOriginalContent');
  if (element instanceof MathfieldPrivate) {
    return element.originalContent;
  }

  // Element is a pair: accessible span, math -- set it to the math part
  element = getElement(element).children[1] as HTMLElement;
  options = options ?? {};
  validateNamespace(options);
  return element.getAttribute(
    'data-' + (options.namespace ?? '') + 'original-content'
  );
}

// This SDK_VERSION variable will be replaced during the build process.
const version = '{{SDK_VERSION}}';

function deprecated(method: string, remedy: string) {
  console.warn(`"${method}" is deprecated. 
${remedy ?? ''}`);
}

function deprecatedDefaultImport(method: string) {
  console.warn(`Using "${method}" as a default import is deprecated.
Instead of
    import Mathlive from 'mathlive';
    ${method}(...);
use
   import ${method} from 'mathlive;
   ${method}(...)  
`);
}

export const debug = {
  getStyle: MathLiveDebug.getStyle,
  getType: MathLiveDebug.getType,
  spanToString: MathLiveDebug.spanToString,
  hasClass: MathLiveDebug.hasClass,
  latexToAsciiMath,
  asciiMathToLatex,
  FUNCTIONS: MathLiveDebug.FUNCTIONS,
  MATH_SYMBOLS: MathLiveDebug.MATH_SYMBOLS,
  TEXT_SYMBOLS: MathLiveDebug.TEXT_SYMBOLS,
  ENVIRONMENTS: MathLiveDebug.ENVIRONMENTS,
  MACROS: MathLiveDebug.MACROS,
  DEFAULT_KEYBINDINGS: MathLiveDebug.DEFAULT_KEYBINDINGS,
  getKeybindingMarkup: MathLiveDebug.getKeybindingMarkup,
};

export default {
  version: (): string => {
    deprecatedDefaultImport('version');
    return version;
  },
  latexToMarkup: (
    text: string,
    options?: {
      mathstyle?: 'displaystyle' | 'textstyle';
      letterShapeStyle?: 'tex' | 'french' | 'iso' | 'upright' | 'auto';
      macros?: MacroDictionary;
      onError?: ErrorListener<ParserErrorCode>;
      format?: string;
    }
  ): string => {
    deprecatedDefaultImport('latexToMarkup');
    return latexToMarkup(text, options);
  },
  latexToMathML: (
    latex: string,
    options?: Partial<{
      macros: MacroDictionary;
      onError: ErrorListener<ParserErrorCode>;
      generateID: boolean;
    }>
  ): string => {
    deprecatedDefaultImport('latexToMathML');
    return latexToMathML(latex, options);
  },
  latexToSpeakableText: (
    latex: string,
    options: Partial<
      TextToSpeechOptions & {
        macros?: MacroDictionary;
        onError?: ErrorListener<ParserErrorCode | MathfieldErrorCode>;
      }
    >
  ): string => {
    deprecatedDefaultImport('latexToSpeakableText');
    return latexToSpeakableText(latex, options);
  },
  latexToAST: (
    latex: string,
    options?: MathJsonLatexOptions & {
      macros?: MacroDictionary;
      onError?: ErrorListener<ParserErrorCode | string>;
    }
  ): string => {
    deprecated('latexToAST', 'Use MathJSON.');
    return latexToAST(latex, options);
  },
  astToLatex: (expr: MathJson, options: MathJsonLatexOptions): string => {
    deprecated('astToLatex', 'Use MathJSON.');
    return astToLatex(expr, options);
  },
  makeMathField: (
    element: HTMLElement,
    options: Partial<MathfieldOptions>
  ): Mathfield => {
    deprecatedDefaultImport('makeMathField');
    return makeMathField(element, options);
  },
  renderMathInDocument: (options?: AutoRenderOptionsPrivate): void => {
    deprecatedDefaultImport('renderMathInDocument');
    renderMathInDocument(options);
  },
  renderMathInElement: (
    element: HTMLElement,
    options: AutoRenderOptionsPrivate
  ): void => {
    deprecatedDefaultImport('renderMathInElement');
    renderMathInElement(element, options);
  },
  revertToOriginalContent: (
    element: string | HTMLElement,
    options: AutoRenderOptionsPrivate
  ): void => {
    deprecatedDefaultImport('revertToOriginalContent');
    revertToOriginalContent(element, options);
  },
  getOriginalContent: (
    element: string | HTMLElement,
    options: AutoRenderOptionsPrivate
  ): void => {
    deprecatedDefaultImport('getOriginalContent');
    getOriginalContent(element, options);
  },

  readAloud: (
    element: HTMLElement,
    text: string,
    config: Partial<MathfieldOptions>
  ): void => {
    deprecatedDefaultImport('readAloud');
    return defaultReadAloudHook(element, text, config);
  },
  readAloudStatus: (): string => {
    deprecatedDefaultImport('readAloudStatus');
    return readAloudStatus();
  },
  pauseReadAloud: (): void => {
    deprecatedDefaultImport('pauseReadAloud');
    pauseReadAloud();
  },
  resumeReadAloud: (): void => {
    deprecatedDefaultImport('resumeReadAloud');
    resumeReadAloud();
  },
  playReadAloud: (token: string, count: number): void => {
    deprecatedDefaultImport('playReadAloud');
    playReadAloud(token, count);
  },
};
