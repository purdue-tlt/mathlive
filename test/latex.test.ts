import { convertLatexToMarkup } from '../dist/mathlive';

function markupAndError(formula: string): [string, string] {
    let errorCode = 'no-error';
    const markup = convertLatexToMarkup(formula, {
        mathstyle: 'displaystyle',
        onError: (err) => {
            if (errorCode === 'no-error') {
                // Catch the first error only
                errorCode = err.code;
            }
        },
    });
    return [markup, errorCode];
}

function error(expression: string) {
    return markupAndError(expression)[1];
}

describe('BASIC PARSING', () => {
    test.each([
        'x',
        ' x ', // Spaces do not matter
        '%', // '%' is start of comment
        '% comment',
        'x % comment',
        'x',
        '-12',
        '1234|/@.`abcdefgzABCDEFGZ', // Basic literals
        'a b', // Spaces are ignored
        'ab', // Same as previous
        'a~b', // ~ is space, same as previous
        'a\\space b',
        '{a}b', // Group
        '{-}', // Operator in group
        '-a', // Spacing as unary operator
        'a-', // Spacing as postfix operator
        'a-b', // Spacing as infix operator
        'a\nb',
        'a=1}',
        'a=1{', // Syntax error
        'a=1{}', // Valid
    ])('%#/ %p renders correctly', (x) => {
        expect(markupAndError(x)).toMatchSnapshot();
    });
    // expect(error('a=1}}}}{{{{')).toMatch('unbalanced-braces');
});

describe('CHARACTERS', () => {
    const ref = convertLatexToMarkup('J0');
    test.each([
        '^^4a0',
        '^^^^004a0',
        '\\char"4A 0',
        "\\char'0112 0",
        '\\char74 0',
        '\\char "004A 0',
        '\\char`J 0',
        '\\char`\\J 0',
        '\\char `\\J 0',
        '\\char   `\\J 0',
        '\\char +- +-  `\\J 0',
        '\\char +- -  `\\J 0',
        '\\char +- -- -++ `\\J 0',
        '\\unicode{"4A} 0',
        '\\unicode{"004A} 0',
        '\\unicode{x004A} 0',
    ])('%#/ %p renders as "J0"', (x) => {
        expect(convertLatexToMarkup(x)).toEqual(ref);
    });
});
describe('EXPANSION PRIMITIVES', () => {
    test.each([
        ['\\obeyspaces =   =', '=\\space\\space\\space='],
        ['\\csname alpha\\endcsname', '\\alpha'],
        ['\\csname alph\\char"41\\endcsname', '\\alph A'],
        ['=\\sqrt\\bgroup x \\egroup=', '=\\sqrt{x}='],
        ['\\string\\alpha', '\\backslash alpha'],
        ['#?', '\\placeholder{}'],
    ])('%#/ %p matches %p', (a, b) => {
        expect(convertLatexToMarkup(a)).toMatch(convertLatexToMarkup(b));
    });
});

describe('ARGUMENTS', () => {
    test.each([
        ['\\frac12', '\\frac{1}{2}'],
        ['\\frac  1  2', '\\frac{1}{2}'],
        ['\\frac357', '\\frac{3}{5}7'],
        ['\\frac3a', '\\frac{3}{a}'],
        ['\\frac\\alpha\\beta', '\\frac{\\alpha}{\\beta}'],
        // ['\\frac{{1}}{2}', '\\frac{1}{2}'],
        ['\\frac  {  { 1  } } { 2 }', '\\frac{{1}}{2}'],
    ])('%#/ %p matches %p', (a, b) => {
        expect(convertLatexToMarkup(a)).toMatch(convertLatexToMarkup(b));
    });
    test.each(['\\frac', '\\frac{}', '\\frac{}{}'])(
        '%#/ %p renders correctly',
        (x) => {
            expect(markupAndError(x)).toMatchSnapshot();
        }
    );
});

describe('INFIX COMMANDS', () => {
    test.each([
        ['a\\over b', '\\frac{a}{b}'],
        ['a\\over b c', '\\frac{a}{bc}'],
        ['x{a+1\\over1-b}y', 'x{\\frac{a+1}{1-b}}y'],
        ['x{a+1\\over1-b\\over2}y', 'x{a+1\\over1-b2}y'],
    ])('%#/ %p matches %p', (a, b) => {
        expect(convertLatexToMarkup(a)).toMatch(convertLatexToMarkup(b));
    });

    expect(error('a\\over b \\over c')).toMatch('too-many-infix-commands');
});
