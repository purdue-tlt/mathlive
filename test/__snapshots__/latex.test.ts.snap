// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ARGUMENTS 0/ "\\\\frac" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:1.2em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:1.89em;vertical-align:-0.68em\\"></span><span class=\\"ML__base\\"><span><span class=\\"sizing nulldelimiter mopen\\"></span><span class=\\"vlist mfrac\\"><span style=\\"top:0.69em\\"></span><span style=\\"top:-0.22em\\"><span class=\\"frac-line\\"></span></span><span style=\\"top:-0.67em\\"></span></span><span class=\\"sizing nulldelimiter mclose\\"></span></span></span></span>",
  "no-error",
]
`;

exports[`ARGUMENTS 1/ "\\\\frac{}" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:1.2em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:1.89em;vertical-align:-0.68em\\"></span><span class=\\"ML__base\\"><span><span class=\\"sizing nulldelimiter mopen\\"></span><span class=\\"vlist mfrac\\"><span style=\\"top:0.69em\\"></span><span style=\\"top:-0.22em\\"><span class=\\"frac-line\\"></span></span><span style=\\"top:-0.67em\\"></span></span><span class=\\"sizing nulldelimiter mclose\\"></span></span></span></span>",
  "no-error",
]
`;

exports[`ARGUMENTS 2/ "\\\\frac{}{}" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:1.2em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:1.89em;vertical-align:-0.68em\\"></span><span class=\\"ML__base\\"><span><span class=\\"sizing nulldelimiter mopen\\"></span><span class=\\"vlist mfrac\\"><span style=\\"top:0.69em\\"></span><span style=\\"top:-0.22em\\"><span class=\\"frac-line\\"></span></span><span style=\\"top:-0.67em\\"></span></span><span class=\\"sizing nulldelimiter mclose\\"></span></span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 0/ "x" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.44em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 1/ " x " renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.44em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 2/ "%" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0em\\"></span><span class=\\"ML__base\\"></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 3/ "% comment" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0em\\"></span><span class=\\"ML__base\\"></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 4/ "x % comment" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.44em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 5/ "x" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.44em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">x</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 6/ "-12" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.65em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:0.73em;vertical-align:-0.08em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__cmr\\">−</span><span class=\\"ML__cmr\\">12</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 7/ "1234|/@.\`abcdefgzABCDEFGZ" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.75em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:1em;vertical-align:-0.25em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__cmr\\">1234∣/@.‘</span><span class=\\"ML__mathit\\">abcdefgzABCDEFGZ</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 8/ "a b" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.7em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 9/ "ab" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.7em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 10/ "a~b" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.7em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 11/ "a\\\\space b" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.7em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 12/ "{a}b" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.7em\\"></span><span class=\\"ML__base\\"><span><span class=\\"ML__mathit\\">a</span></span><span class=\\"ML__mathit\\">b</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 13/ "{-}" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.59em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:0.67em;vertical-align:-0.08em\\"></span><span class=\\"ML__base\\"><span><span class=\\"ML__cmr\\">−</span></span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 14/ "-a" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.59em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:0.67em;vertical-align:-0.08em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__cmr\\">−</span><span class=\\"ML__mathit\\">a</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 15/ "a-" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.59em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:0.67em;vertical-align:-0.08em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">a</span><span class=\\"ML__cmr\\">−</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 16/ "a-b" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.7em\\"></span><span class=\\"ML__strut--bottom\\" style=\\"height:0.78em;vertical-align:-0.08em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">a</span><span class=\\"ML__cmr\\" style=\\"margin-left:0.23em\\">−</span><span class=\\"ML__mathit\\" style=\\"margin-left:0.23em\\">b</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 17/ "a
b" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.7em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">ab</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 18/ "a=1}" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.65em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">a</span><span class=\\"ML__cmr\\" style=\\"margin-left:0.28em\\">=</span><span class=\\"ML__cmr\\" style=\\"margin-left:0.28em\\">1</span></span></span>",
  "no-error",
]
`;

exports[`BASIC PARSING 19/ "a=1{" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.65em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">a</span><span class=\\"ML__cmr\\" style=\\"margin-left:0.28em\\">=</span><span class=\\"ML__cmr\\" style=\\"margin-left:0.28em\\">1</span></span></span>",
  "unbalanced-braces",
]
`;

exports[`BASIC PARSING 20/ "a=1{}" renders correctly 1`] = `
Array [
  "<span class=\\"ML__mathlive\\"><span class=\\"ML__strut\\" style=\\"height:0.65em\\"></span><span class=\\"ML__base\\"><span class=\\"ML__mathit\\">a</span><span class=\\"ML__cmr\\" style=\\"margin-left:0.28em\\">=</span><span class=\\"ML__cmr\\" style=\\"margin-left:0.28em\\">1</span></span></span>",
  "no-error",
]
`;
